### PR TITLE
6 export types w/ up to 2 [_] as initial type members

### PR DESCRIPTION
**Pull request by wheaties**
A little boilerplate and a change from quasiquotes to direct trees. If this is good, I'll work on the `imports`.

### CODE REVIEW

**Comment by milessabin** [comment]
I like the direction this is going in. I can't see the 2.10.x problem immediately ... I suggest you extract out the generated source and try compiling it directly as ordinary source.

**Comment by milessabin** [comment]
It looks like I need support for type classes which take a pair of type constructors (ie. `TC[F[_], G[_]]`) or "11" in the numeric encoding ... does your boilerplate generation generalize to that as well?

**Comment by wheaties** [comment]
No, it only goes up to 10. I will have to play around with getting 2.10 to compile. The non-generated code should be able to handle any variant of 1 and 0.

**Comment by wheaties** [comment]
I was able to half get the quasi-quote working but ran into issues surrounding anything that used an imports or reexports in the 2.10.6 branch. I've switched back to manually writing out the full arity expansion for the exports.

File: project/Boilerplate.scala [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    Blocks in the templates below use a custom interpolator, combined with post-processing to produce the body
+      - The contents of the `header` val is output first
+      - Then the first block of lines beginning with '|'
+      - Then the block of lines beginning with '-' is replicated once for each arity,
+        with the `templateVals` already pre-populated with relevant relevant vals for that arity
+      - Then the last block of lines prefixed with '|'
+    The block otherwise behaves as a standard interpolated string with regards to variable substitution.
+  */
+
+  object ExportsImpl extends Template{
```
Comment (milessabin) [inline project/Boilerplate.scala:97]: There's quite a lot here which is independent of the interpolation variables ... I'd prefer it if you pulled that out into non-generated code.

File: project/Boilerplate.scala [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        |    val exportTc = eTpe.typeConstructor
+        |    val exportTpe = appliedType(exportTc, appliedType(tcTpe, tTpes))
+        |    q"new $$exportTpe($$st)"
+        |  }
+        |}
+      """
+    }
+  }
+
+  object ImportsImpl extends Template{
```
Comment (milessabin) [inline project/Boilerplate.scala:123]: Same here ...

File: build.sbt [hunk 3]
```diff
@@ -137,12 +138,32 @@ lazy val mimaSettings = mimaDefaultSettings ++ Seq(
     // Filtering the methods that were added since the checked version
     // (these only break forward compatibility, not the backward one)
     Seq(
-      exclude[MissingMethodProblem]("export.ExportMacro#Stub.exportsImpl10"),
```
Comment (milessabin) [inline build.sbt:140]: Just to be sure ... are all these "missing in the old version"?

Comment (wheaties) [inline build.sbt:140]: Yes. I just moved it so that it reads sequentially.

File: project/Boilerplate.scala (hunk 1 above)
Comment (wheaties) [inline project/Boilerplate.scala:97]: Ok, so should I move the `importsImplAux` back into the `ExportMacro` class and in doing so, am I ok to assume that the `c.universe._` will be the same variable and instance? If so, I'll get on that tonight (ET, US.)

**Comment by milessabin** [comment]
LGTM ... Many thanks! Merging :-)

**Pull request merged**